library/ssl_msg.c

```diff
@@ -5865,54 +5865,20 @@ int mbedtls_ssl_read(mbedtls_ssl_context *ssl, unsigned char *buf, size_t len)
     return ret;
 }
 
-
 #if defined(MBEDTLS_SSL_SRV_C) && defined(MBEDTLS_SSL_EARLY_DATA)
 int mbedtls_ssl_read_early_data(mbedtls_ssl_context *ssl,
                                 unsigned char *buf, size_t len)
 {
-    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    const struct mbedtls_ssl_config *conf;
-    unsigned char *p = buf;
-
-    if (ssl == NULL || ((conf = ssl->conf) == NULL)) {
+    if (ssl == NULL || (ssl->conf == NULL)) {
         return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
     }
 
-    if ((!mbedtls_ssl_conf_is_tls13_enabled(conf)) ||
-        (conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM) ||
-        (conf->early_data_enabled != MBEDTLS_SSL_EARLY_DATA_ENABLED)) {
-        return MBEDTLS_ERR_SSL_CANNOT_READ_EARLY_DATA;
-    }
-
-    if (ssl->tls_version != MBEDTLS_SSL_VERSION_TLS1_3) {
-        return MBEDTLS_ERR_SSL_CANNOT_READ_EARLY_DATA;
-    }
-
-    if ((ssl->early_data_state.srv !=
-         MBEDTLS_SSL_SRV_EARLY_DATA_STATE_WAITING_CH) &&
-        (ssl->early_data_state.srv !=
-         MBEDTLS_SSL_SRV_EARLY_DATA_STATE_ACCEPTING)) {
-        return MBEDTLS_ERR_SSL_CANNOT_READ_EARLY_DATA;
-    }
-
-    ret = mbedtls_ssl_handshake(ssl);
-    if (ret == MBEDTLS_ERR_SSL_RECEIVED_EARLY_DATA) {
-        if (ssl->in_offt == NULL) {
-            /* Set the reading pointer */
-            ssl->in_offt = ssl->in_msg;
-        }
-        ret = ssl_read_application_data(ssl, p, len);
-    } else if (ret == 0) {
-        /*
-         * If the handshake is completed, return immediately that early data
-         * cannot be read anymore. This potentially saves another call to this
-         * API and when the function returns 0, it only means that zero byte
-         * of early data has been received.
-         */
+    if ((ssl->state != MBEDTLS_SSL_END_OF_EARLY_DATA) ||
```

Comment:
The state's name makes this a bit counter-intuitive. IIRC states are name after the next message we're expecting, but in that case we're expecting either early data or end early data. Should rename the state to reflect that? Like `MBEDTLS_SSL_EARLY_DATA_OR_END_OF_EARLY_DATA`?

Also, I'm surprised that we apparently don't need to check `ssl->early_data_state.srv` at all?

Reply:
> The state name is used on client side as well where MBEDTLS_SSL_EARLY_DATA_OR_END_OF_EARLY_DATA would not be suitable.

Ah, good point. Thanks for adding a comment.

Reply:
> Now we just need a flag to note down if the server has accepted early data or not at reception of the ClientHello. I've reworked that.

Nice, looks much simpler now. Just to make sure I understand: we don't need to check this flag here because the only way `ssl->state` can be equal to `MBEDTLS_SSL_END_OF_EARLY_DATA` is if the server has accepted early data?

Also, does `ssl->in_offt != NULL` guarantee that the server has just received early data? If that's the case that's really not obvious to me - I though `in_offt` was also used for other purposes (for example multiple handshake messages in a single record or things like that).

To put my question in context, what I'm worried about is the following scenario:
1. Application calls `ssl_handhsake()` (or even `handshake_step()`).
2. `ssl_handshake()` returns something _other_ than `RECEIVED_EARLY_DATA`.
3. Application call `mbedtls_ssl_read_early_data()` anyway because misusing the API is what users do.
4. Do (a) we detect the misuse and cleanly return an error, or (b) silently output garbage?

Right now it's not clear to me that we're doing (a) not (b) - at least not in all cases. Wdyt?

Reply:
> Nice, looks much simpler now. Just to make sure I understand: we don't need to check this flag here because the only way `ssl->state` can be equal to `MBEDTLS_SSL_END_OF_EARLY_DATA` is if the server has accepted early data?

Yes that's right, see lines 2867 to 2876. That's the only code on server side to move to the `MBEDTLS_SSL_END_OF_EARLY_DATA` state.

Reply:
> I though `in_offt` was also used for other purposes (for example multiple handshake messages in a single record or things like that).

It seems to me that `in_offt` is dedicated to application data.
Declaration in `ssl.h`:
```C
unsigned char *MBEDTLS_PRIVATE(in_offt);     /*!< read offset in application data  */
```
and looking its usage in ssl_msg.c it seems to be indeed the case.

Reply:
Ah right, it seems my memories were blurry and I was mixing up different things in my mind (like `in_left`, `next_record_offset` etc.). Indeed this is only for application data, so it being non-NULL guarantees that we indeed received early/application data. Thanks for clarifying!